### PR TITLE
CIRC-1046 add handling for not critical errors for check-out

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
Check out should respond with 200 in case of partially successful checkout. For full successful checkout, the status code should not be changed.

## Approach
Implement error handling for all actions that follow after creating a loan in the storage. After creating it, an item status should be "checked out" and these not critical issues should be handled with 200 status.

Resolves: [CIRC-1046](https://issues.folio.org/browse/CIRC-1046)

### CODE REVIEW

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 1]
```diff
@@ -1443,21 +1445,66 @@ void checkOutRefusedWhenAutomatedBlockExistsForPatron() {
...
+        .to(steve)
+        .on(getZonedDateTime())
+        .at(UUID.randomUUID()));
+    FakePubSub.setFailPublishingWithBadRequestError(false);
+
+    assertThat(itemsClient.get(smallAngryPlanet), hasItemStatus(CHECKED_OUT));
+  }
+
+  @Test
+  void checkOutShouldNotFailIfSessionRecordCreatingFails() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1465]:
```suggestion
  void checkOutShouldNotFailIfSessionRecordCreationFails() {
```

File: src/test/java/api/support/fakes/FakeStorageModule.java [hunk 2]
```diff
@@ -51,6 +51,10 @@
 public class FakeStorageModule extends AbstractVerticle {
   private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
   private static final Set<String> queries = Collections.synchronizedSet(new HashSet<>());
+  public static final String PATRON_ACTION_SESSION_STORAGE_PATH =
+    "/patron-action-session-storage/patron-action-sessions";
+  public static final String SCHEDULED_NOTICE_STORAGE_PATH =
+    "/scheduled-notice-storage/scheduled-notices";
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/support/fakes/FakeStorageModule.java:57]: Do you think it's a good idea to hardcode the knowledge of specific endpoints into this class? Should't only specific instances of this class know that? I know that we don't have access to instances of `FakeStorageModule` from the tests, and I understand that we need to make these instances return errors in tests sometimes, but can we think of a more generic way to achieve that using static methods of `FakeStorageModule`? Something like:
```
FakeStorageModule.fail(HttpMethod method, String url, HttpStatus statusCode) {...}
```

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 3]
```diff
@@ -1443,21 +1445,66 @@ void checkOutRefusedWhenAutomatedBlockExistsForPatron() {
   }
 
   @Test
-  void checkOutFailsWhenEventPublishingFailsWithBadRequestError() {
+  void checkOutShouldNotFailIfEventPublishingFailsWithBadRequestError() {
```
Comment (sherzodn) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1448]: do we have a test case that returns a 201 status code?

Comment (roman-barannyk) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java:1453]: Sure, the most of the tests use checkOutFixture.checkOutByBarcode which checks inside 201 status

File: src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeService.java [hunk 4]
```diff
@@ -126,7 +132,7 @@ private ScheduledNoticeConfig createScheduledNoticeConfig(NoticeConfiguration co
 
     if (!loan.isClosed()) {
       scheduledNoticesRepository.deleteByLoanIdAndTriggeringEvent(loan.getId(), triggeringEvent)
-        .thenAccept(r -> r.next(v -> scheduleLoanNotices(loan, eventType, eventTime)));
+        .thenAccept(r -> r.map(v -> scheduleLoanNotices(loan, eventType, eventTime)));
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/LoanScheduledNoticeService.java]: done

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java [hunk 5]
```diff
@@ -173,8 +206,13 @@ private LoanAndRelatedRecords checkOutItem(LoanAndRelatedRecords loanAndRelatedR
     return loanAndRelatedRecords.changeItemStatus(CHECKED_OUT);
   }
 
-  private Result<HttpResponse> createdLoanFrom(Result<JsonObject> result) {
-    return result.map(json -> created(json, urlForLoan(json.getString("id"))));
+  private Result<HttpResponse> createdLoanFrom(Result<JsonObject> result,
+    CirculationErrorHandler errorHandler) {
+
+    return result.map(json -> errorHandler.hasAny(FAILED_TO_SAVE_SESSION_RECORD,
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java]: extracted

File: src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java [hunk 6]
```diff
@@ -173,8 +206,13 @@ private LoanAndRelatedRecords checkOutItem(LoanAndRelatedRecords loanAndRelatedR
...
+      FAILED_TO_SCHEDULE_NOTICE_FOR_LOAN_DUE_DATE, FAILED_TO_PUBLISH_CHECKOUT_EVENT)
+        ? ok(json, urlForLoan(json.getString("id")))
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java]: extracted

File: src/main/java/org/folio/circulation/resources/handlers/error/CirculationErrorType.java [hunk 7]
```diff
@@ -53,5 +53,10 @@
   RENEWAL_IS_NOT_POSSIBLE,
 
   // Errors related to block overrides
-  INSUFFICIENT_OVERRIDE_PERMISSIONS
+  INSUFFICIENT_OVERRIDE_PERMISSIONS,
+
+  // Errors that do not affect a loan creation
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/resources/handlers/error/CirculationErrorType.java]: updated

File: src/test/java/api/loans/CheckOutByBarcodeTests.java (hunk 1 above)
Comment (roman-barannyk) [reply src/test/java/api/loans/CheckOutByBarcodeTests.java]: renamed

File: src/test/java/api/support/fakes/FakeStorageModule.java (hunk 2 above)
Comment (roman-barannyk) [reply src/test/java/api/support/fakes/FakeStorageModule.java]: upated, thanks

**Review by OleksandrVidinieiev — APPROVED** [review]

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**